File: O SIA Programación/SIAProgramación.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace SistemaInmuneArtificial
{
    public static class Utils
    {
        // Convierte un BitArray en una cadena de caracteres representando 1s y 0s
        public static string BitArrayAsString(BitArray ba)
        {
            string result = "";
            foreach (bool bit in ba)
            {
                result += bit ? "1" : "0";
            }
            return result;
        }
    }

    class ProgramaSistemaInmuneArtificial
    {
        static Random random = new Random();

        static void Main(string[] args)
        {
            Console.WriteLine("\nComienzo de la demostración del Sistema Inmune Artificial para la Detección de Intrusos\n");

            int numPatternBits = 12;
            int numAntibodyBits = 4;
            int numLymphocytes = 3;
            int stimulationThreshold = 3;

            Console.WriteLine("Cargando conjunto de autoantígenos ('patrones históricos normales)");
            List<BitArray> selfSet = LoadSelfSet();
            ShowSelfSet(selfSet);

            Console.WriteLine("\nCreando conjunto de linfocitos utilizando selección negativa y detección de r-chunks");
            List<Lymphocyte> lymphocyteSet = CreateLymphocyteSet(selfSet, numAntibodyBits, numLymphocytes);
            ShowLymphocyteSet(lymphocyteSet);

            Console.WriteLine("\nComienzo de la simulación de detección de intrusiones del AIS\n");
            int time = 0;
            int maxTime = 6;
            List<BitArray> detectedAndRemovedPatterns = new List<BitArray>();
            int intrusionCount = 0;

            while (time < maxTime)
            {
                Console.WriteLine("============================================");
                BitArray incoming = RandomBitArray(numPatternBits);
                Console.WriteLine("Patrón entrante = " + Utils.BitArrayAsString(incoming) + "\n");

				bool isDetected = false;
                foreach (var lymph in lymphocyteSet)
                {
                    if (lymph.Detects(incoming))
                    {
                        Console.WriteLine("Patrón entrante detectado por un linfocito");
                        lymph.Stimulation++;
                        if (lymph.Stimulation >= stimulationThreshold)
                        {
                            Console.WriteLine("¡Linfocito estimulado! Patrón marcado para eliminación como posible intrusión.");
                            if (!isDetected)
                            {
                                detectedAndRemovedPatterns.Add(incoming);
                                intrusionCount++;
                                isDetected = true;
                            }
                        }
                        else
                        {
                            Console.WriteLine("Linfocito no supera el umbral de estimulación");
                        }
                    }
                    else
                    {
                        Console.WriteLine("Patrón entrante no detectado por el linfocito");
                    }
                }
                ++time;
                Console.WriteLine("============================================");
            }

            Console.WriteLine("\nIntrusiones detectadas:");
            foreach (var intrusion in detectedAndRemovedPatterns)
            {
                Console.WriteLine(Utils.BitArrayAsString(intrusion));
            }
            Console.WriteLine($"Total de intrusiones detectadas: {intrusionCount}");

            Console.WriteLine("\nFin de la demo del IDS AIS\n");
            Console.ReadLine();
        }

        static List<BitArray> LoadSelfSet()
        {
            // Cargar set de auto-antígenos, representando patrones normales
            return new List<BitArray>();
        }

        static void ShowSelfSet(List<BitArray> selfSet)
        {
            // Mostrar el set de auto-antígenos
            foreach (var ba in selfSet)
            {
                Console.WriteLine(Utils.BitArrayAsString(ba));
            }
        }

        static List<Lymphocyte> CreateLymphocyteSet(List<BitArray> selfSet, int numAntibodyBits, int numLymphocytes)
        {
            // Crear un set de linfocitos usando selección negativa y detección r-chunks
            return new List<Lymphocyte>();
        }

        static void ShowLymphocyteSet(List<Lymphocyte> lymphocyteSet)
        {
            // Mostrar el set de linfocitos
            Console.WriteLine("\nLymphocyte set:");
            foreach (Lymphocyte lymph in lymphocyteSet)
            {
                Console.WriteLine(lymph.ToString());
            }
        }

        static BitArray RandomBitArray(int numBits)
        {
            // Generar un BitArray aleatorio para simular patrones entrantes
            BitArray ba = new BitArray(numBits);
            for (int i = 0; i < numBits; ++i)
            {
                ba[i] = random.NextDouble() > 0.5;
            }
            return ba;
        }
    }

    class Lymphocyte
    {
        public BitArray Antibody;
        public int Stimulation;

        public Lymphocyte(BitArray antibody)
        {
            Antibody = antibody;
            Stimulation = 0;
        }

        // Detectar si un patrón coincide con el anticuerpo del linfocito
        public bool Detects(BitArray pattern)
        {
            for (int i = 0; i < pattern.Count; ++i)
            {
                if (pattern[i] != Antibody[i])
                    return false;
            }
            return true;
        }

        public override string ToString()
        {
            // Representación en cadena de un linfocito
            return $"Antibody: {Utils.BitArrayAsString(Antibody)}, Stimulation: {Stimulation}";
        }
    }
}

File: O SIA Programación/SIAProgramación.py
from random import random
from typing import List


class Utils:
    @staticmethod
    def bit_array_as_string(bit_array):
        return ''.join(['1' if bit else '0' for bit in bit_array])


class Lymphocyte:
    def __init__(self, antibody):
        self.antibody = antibody
        self.stimulation = 0

    def detects(self, pattern):
        return all(x == y for x, y in zip(self.antibody, pattern))

    def __str__(self):
        return f"Antibody: {Utils.bit_array_as_string(self.antibody)}, Stimulation: {self.stimulation}"


def random_bit_array(num_bits):
    return [random() > 0.5 for _ in range(num_bits)]


def load_self_set():
    return []


def show_self_set(self_set):
    for ba in self_set:
        print(Utils.bit_array_as_string(ba))


def create_lymphocyte_set(self_set, num_antibody_bits, num_lymphocytes):
    return []


def show_lymphocyte_set(lymphocyte_set):
    print("\nLymphocyte set:")
    for lymph in lymphocyte_set:
        print(lymph)


def main():
    print("\nComienzo de la demostración del Sistema Inmune Artificial para la Detección de Intrusos\n")

    num_pattern_bits = 12
    num_antibody_bits = 4
    num_lymphocytes = 3
    stimulation_threshold = 3

    print("Cargando conjunto de autoantígenos ('patrones históricos normales')")
    self_set = load_self_set()
    show_self_set(self_set)

    print("\nCreando conjunto de linfocitos utilizando selección negativa y detección de r-chunks")
    lymphocyte_set = create_lymphocyte_set(self_set, num_antibody_bits, num_lymphocytes)
    show_lymphocyte_set(lymphocyte_set)

    print("\nComienzo de la simulación de detección de intrusiones del AIS\n")
    time = 0
    max_time = 6
    detected_and_removed_patterns = []
    intrusion_count = 0

    while time < max_time:
        print("============================================")
        incoming = random_bit_array(num_pattern_bits)
        print(f"Patrón entrante = {Utils.bit_array_as_string(incoming)}\n")

        is_detected = False
        for lymph in lymphocyte_set:
            if lymph.detects(incoming):
                print("Patrón entrante detectado por un linfocito")
                lymph.stimulation += 1
                if lymph.stimulation >= stimulation_threshold:
                    print("¡Linfocito estimulado! Patrón marcado para eliminación como posible intrusión.")
                    if not is_detected:
                        detected_and_removed_patterns.append(incoming)
                        intrusion_count += 1
                        is_detected = True
                else:
                    print("Linfocito no supera el umbral de estimulación")
            else:
                print("Patrón entrante no detectado por el linfocito")
        time += 1
        print("============================================")

    print("\nIntrusiones detectadas:")
    for intrusion in detected_and_removed_patterns:
        print(Utils.bit_array_as_string(intrusion))
    print(f"Total de intrusiones detectadas: {intrusion_count}")

    print("\nFin de la demo del IDS AIS\n")


if __name__ == '__main__':
    main()

